winsup/utils/getprocaddr.c

```diff
@@ -0,0 +1,198 @@
+#include <stdio.h>
+#include <windows.h>
+
+/**
+ * To determine the address of kernel32!CtrlRoutine, we need to use
+ * dbghelp.dll. But we want to avoid linking statically to that library because
+ * the normal operation of cygwin-console-helper.exe (i.e. to allocate a hidden
+ * Console) does not need it.
+ *
+ * Therefore, we declare the SYMBOL_INFOW structure here, load the dbghelp
+ * library via LoadLibraryExA() and obtain the SymInitialize(), SymFromAddrW()
+ * and SymCleanup() functions via GetProcAddr().
+ * 
+ * After that a remote thread is injected into that process
+ * to send the exit code.
+ */
+
+#include <dbghelp.h>
+
+static int pid;
+static uintptr_t exit_code;
+static BOOL do_print_address;
+
+/* Avoid fprintf(), as it would try to reference '__getreent' */
+static void
+output (BOOL error, const char *fmt, ...)
+{
+  va_list ap;
+  char buffer[1024];
+
+  va_start (ap, fmt);
+  vsnprintf (buffer, sizeof(buffer) - 1, fmt, ap);
+  buffer[sizeof(buffer) - 1] = '\0';
+  va_end (ap);
+  WriteFile (GetStdHandle(error ? STD_ERROR_HANDLE : STD_OUTPUT_HANDLE),
+	      buffer, strlen (buffer), NULL, NULL);
+}
+
+static int
+inject_remote_thread_into_process(HANDLE process, LPTHREAD_START_ROUTINE address, uintptr_t exit_code)
+{
+  int res = -1;
+
+  if (!address)
+    return res;
+  DWORD thread_id;
+  HANDLE thread = CreateRemoteThread(process, NULL, 1024 * 1024, address,
+                                     (PVOID)exit_code, 0, &thread_id);
+  if (thread)
+  {
+    /*
+      * Wait up to 10 seconds (arbitrary constant) for the thread to finish;
+      * After that grace period, fall back to terminating non-gently.
+      */
+    if (WaitForSingleObject(thread, 10000) == WAIT_OBJECT_0)
+      res = 0;
+    CloseHandle(thread);
+  }
+
+  return res;
+}
+
+static WINAPI BOOL
+ctrl_handler(DWORD ctrl_type)
```

Comment:
I think this function makes for a fine spot to add a multi-line comment describing the strategy where we send a CtrlEvent to the current process for the _sole_ purpose of capturing the address of the `CtrlRoutine` function, by looking at the stack trace. This description should stress that we would _love_ to use `GetProcAddress()` as we do for `ExitProcess()`, but sadly `CtrlRoutine` is not exported (although the `.pdb` files ensure that we can see it in a debugger).

Reply:
Done in https://github.com/msys2/msys2-runtime/pull/31/commits/9489d1ad2401b33cf1b9d881329fbb2a76f4945c